contracts/svm/programs/express_relay/src/lib.rs

```diff
@@ -309,3 +398,119 @@ pub struct WithdrawFees<'info> {
     #[account(mut, seeds = [SEED_METADATA], bump, has_one = admin)]
     pub express_relay_metadata: Account<'info, ExpressRelayMetadata>,
 }
+
+/// For all swap instructions and contexts, input and output are defined with respect to the searcher
+/// So mint_input refers to the token that the searcher provides to the trader
+/// mint_output refers to the token that the searcher receives from the trader
+/// This choice is made to minimize confusion for the searchers, who are more likely to parse the program
+#[derive(AnchorSerialize, AnchorDeserialize, Eq, PartialEq, Clone, Copy, Debug)]
+pub struct SwapArgs {
+    pub amount_input:       u64,
+    pub amount_output:      u64,
+    pub nonce:              u64,
+    // The referral fee is specified in parts per million (e.g. 100 = 1 bp)
+    pub referral_fee_ppm:   u64,
+    // Whether the referral fee should be applied to the input token
+    pub referral_fee_input: bool,
+}
+
+#[derive(Accounts)]
+#[instruction(data: Box<SwapArgs>)]
+pub struct Swap<'info> {
+    #[account(mut)]
+    pub searcher: Signer<'info>,
+
+    pub trader: Signer<'info>,
+
+    pub relayer_signer: Signer<'info>,
+
+    #[account(seeds = [
+        SEED_SWAP,
+        searcher.key().as_ref(),
+        trader.key().as_ref(),
+        mint_input.key().as_ref(),
+        mint_output.key().as_ref(),
+        &data.nonce.to_le_bytes(),
+    ], bump)]
+    pub permission: AccountInfo<'info>,
+
+    /// CHECK: don't care what this looks like
+    #[account(mut)]
+    pub router: UncheckedAccount<'info>,
+
+    /// CHECK: this cannot be checked against ConfigRouter bc it may not be initialized bc anchor. we need to check this config even when unused to make sure unique fee splits don't exist
+    #[account(seeds = [SEED_CONFIG_ROUTER, router.key().as_ref()], bump)]
+    pub config_router: UncheckedAccount<'info>,
+
+    #[account(mut, seeds = [SEED_METADATA], bump, has_one = relayer_signer)]
+    pub express_relay_metadata: Account<'info, ExpressRelayMetadata>,
+
+    #[account(mint::token_program = token_program_input)]
+    pub mint_input: InterfaceAccount<'info, Mint>,
+
+    #[account(mint::token_program = token_program_output)]
+    pub mint_output: InterfaceAccount<'info, Mint>,
+
+    #[account(
+        mut,
+        token::mint = mint_input,
+        token::authority = searcher,
+        token::token_program = token_program_input
+    )]
+    pub ta_input_searcher: InterfaceAccount<'info, TokenAccount>,
+
+    #[account(
+        mut,
+        token::mint = mint_output,
+        token::authority = searcher,
+        token::token_program = token_program_output
+    )]
+    pub ta_output_searcher: InterfaceAccount<'info, TokenAccount>,
```

Comment:
is it possible that the searcher doesn't have an output ata yet?

Reply:
possible, although initializing this would need to be added into the logic of the method. also this wouldn't be straightforward to initialize if we wanted to keep it generic to all token accounts (not just ATAs)

Reply:
mmmmmmmmmm, why is ok to have init_if_needed for the router token accounts but not here